README.md
[![Build Status][ci-badge]][ci-badge-url]
[![pub package](https://img.shields.io/pub/v/naked_mapper.svg)](https://pub.dartlang.org/packages/naked_mapper)

This package allows programmers to annotate Dart objects in order to
  Serialize / Deserialize them to / from JSON.

## Why?

* Compatible with **all** target platforms for Dart language, **NO** dependency on `dart:mirrors`, one of the reasons is described [here][1].
* No need to extend your classes from **any** mixins/base/abstract classes to keep code leaner
* No enforced private constructors, do not require any constructors at all for your data-only classes
* No magic `_$` prefixes, No enforced `static` fields in your classes
* Clean and simple setup, transparent and straight-forward usage with **no heavy maintenance**
* Inspired by [json2typescript][4], [serde][5], [gson][6], feature parity with highly popular [Java Jackson][12] and only **4** [annotations](#annotations) to remember to cover all possible use cases.
* **No extra boilerplate**, no messy extra `*.g.dart` files per *each* meaningful file (single root-level file which contains all of the generated code)
* **Complementary adapters** full control over the process when you strive for maximum flexibility.
* `Configuration over code` brings **predictability** to your codebase, while reducing overall amount of code to read / maintain
* Because Serialization/Deserialization is **NOT** a responsibility of your Model classes.

Dart classes reflection mechanism is based on [reflectable][3] library.
This means "extended types information" is auto-generated out of existing Dart program
guided by the annotated classes **only**, as the result types information is accessible at runtime, at a reduced cost.


![](banner.svg)

* [Basic setup](#basic-setup)
* [Annotations](#annotations)
* [Builder](#builder)
* [Known limitations](#known-limitations)
* [Documentation][docs]
* [Configuration use cases](#format-datetime--num-types)
    * [Extended classes](#inherited-classes-derived-from-abstract--base-class)
    * [Classes with Mixins](#classes-enhanced-with-mixins-derived-from-abstract-class)
    * [Immutable classes](#example-with-immutable-class)
    * [Get or Set fields](#get-or-set-fields)
    * [Constructor parameters](#constructor-parameters)
    * [Unmapped properties](#unmapped-properties)
    * [DateTime / num types](#format-datetime--num-types)
    * [Iterable types](#iterable-types)
    * [Value injection](#value-injection)
    * [Enum types](#enum-types)
    * [Enums having String / num values](#enums-having-string--num-values)
    * [Name casing styles](#name-casing-styles-pascal-kebab-snake-snakeallcaps)
    * [Serialization template](#serialization-template)
    * [Deserialization template](#deserialization-template)
    * [Custom types](#custom-types)
    * [Nesting](#nesting-configuration)
    * [Name aliases](#name-aliases-configuration)
    * [Relative path reference to parent field from nested object "../id"](#relative-path-reference-to-parent-field-from-nested-object-id)
    * [Relative path reference to parent itself from nested object ".."](#relative-path-reference-to-parent-itself-from-nested-object-)
    * [Schemes](#schemes)
    * [Objects flattening](#objects-flattening)
    * [Objects cloning](#objects-cloning)
* [Adapters](#complementary-adapter-libraries)
    * [How to use adapter?](#complementary-adapter-libraries)
    * [![pub package](https://img.shields.io/pub/v/dart_json_mapper_built.svg)](https://pub.dartlang.org/packages/dart_json_mapper_built) | [dart_json_mapper_built](adapters/built) | [Built Collection][16]
    * [![pub package](https://img.shields.io/pub/v/dart_json_mapper_mobx.svg)](https://pub.dartlang.org/packages/dart_json_mapper_mobx) | [dart_json_mapper_mobx](adapters/mobx) | [MobX][7]
    * [![pub package](https://img.shields.io/pub/v/dart_json_mapper_fixnum.svg)](https://pub.dartlang.org/packages/dart_json_mapper_fixnum) | [dart_json_mapper_fixnum](adapters/fixnum) | [Fixnum][8]
    * [![pub package](https://img.shields.io/pub/v/dart_json_mapper_flutter.svg)](https://pub.dartlang.org/packages/dart_json_mapper_flutter) | [dart_json_mapper_flutter](adapters/flutter) | [Flutter][11]

## Basic setup

Please add the following dependencies to your `pubspec.yaml`:

```yaml
dependencies:
  naked_mapper:
dev_dependencies:
  build_runner:
```

Say, you have a dart program *main.dart* having some classes intended to be traveling to JSON and back.
- First thing you should do is to put `@jsonSerializable` annotation on each of those classes
- Next step is to auto generate *main.mapper.g.dart* file. And afterwards import that file into *main.dart*

**lib/main.dart**
```dart
import 'package:naked_mapper/dart_json_mapper.dart' show JsonMapper, jsonSerializable, JsonProperty;

import 'main.mapper.g.dart' show initializeJsonMapper;

@jsonSerializable // This annotation let instances of MyData travel to/from JSON
class MyData {
  int a = 123;

  @JsonProperty(ignore: true)
  bool b;

  @JsonProperty(name: 'd')
  String c;

  MyData(this.a, this.b, this.c);
}

void main() {
  initializeJsonMapper();

  print(JsonMapper.serialize(MyData(456, true, "yes")));
}
```
output:
```json
{
  "a": 456,
  "d": "yes"
}
```

Go ahead and create / update `build.yaml` file in your project root directory with the following snippet:

```yaml
targets:
  $default:
    builders:
      naked_mapper:
        generate_for:
        # here should be listed entry point files having 'void main()' function
          - lib/main.dart

      # This part is needed to tell original reflectable builder to stay away
      # it overrides default options for reflectable builder to an **empty** set of files
      reflectable:
        generate_for:
          - no/files
```

Now run the code generation step with the root of your package as the current directory:

```shell
> pub run build_runner build --delete-conflicting-outputs
```

**You'll need to re-run code generation each time you are making changes to `lib/main.dart`**
So for development time, use `watch` like this

```shell
> pub run build_runner watch --delete-conflicting-outputs
```

Each time you modify your project code, all `*.mapper.g.dart` files will be updated as well.
- Next step is to add `*.mapper.g.dart` to your .gitignore
- And this is it, you are all set and ready to go. Happy coding!

## Format DateTime / num types

In order to format `DateTime` or `num` instance as a JSON string, it is possible to
provide [intl][2] based formatting patterns.

**DateTime**
```dart
@JsonProperty(converterParams: {'format': 'MM-dd-yyyy H:m:s'})
DateTime lastPromotionDate = DateTime(2008, 05, 13, 22, 33, 44);

@JsonProperty(converterParams: {'format': 'MM/dd/yyyy'})
DateTime hireDate = DateTime(2003, 02, 28);
```
output:
```json
{
"lastPromotionDate": "05-13-2008 22:33:44",
"hireDate": "02/28/2003"
}
```

**num**
```dart
@JsonProperty(converterParams: {'format': '##.##'})
num salary = 1200000.246;
```
output:
```json
{
"salary": "1200000.25"
}
```

As well, it is possible to utilize `converterParams` map to provide custom
parameters to your [custom converters](#custom-types).

## Get or Set fields

When relying on Dart `getters / setters`, no need to annotate them.
But when you have custom `getter / setter` methods, you should provide annotations for them.

```dart
@jsonSerializable
class AllPrivateFields {
  String _name;
  String _lastName;

  set name(dynamic value) {
    _name = value;
  }

  String get name => _name;

  @JsonProperty(name: 'lastName')
  void setLastName(dynamic value) {
    _lastName = value;
  }

  @JsonProperty(name: 'lastName')
  String getLastName() => _lastName;
}

// given
final json = '''{"name":"Bob","lastName":"Marley"}''';

// when
final instance = JsonMapper.deserialize<AllPrivateFields>(json);

// then
expect(instance.name, 'Bob');
expect(instance.getLastName(), 'Marley');

// when
final targetJson = JsonMapper.serialize(instance, SerializationOptions(indent: ''));

// then
expect(targetJson, json);
```

## Example with immutable class

```dart
@jsonSerializable
enum Color { red, blue, green, brown, yellow, black, white }

@jsonSerializable
class Car {
    @JsonProperty(name: 'modelName')
    String model;

    Color color;

    @JsonProperty(ignore: true)
    Car replacement;

    Car(this.model, this.color);
}

@jsonSerializable
class Immutable {
    final int id;
    final String name;
    final Car car;

    const Immutable(this.id, this.name, this.car);
}

print(
  JsonMapper.serialize(
    Immutable(1, 'Bob', Car('Audi', Color.green))
  )
);
```
output:
```json
{
 "id": 1,
 "name": "Bob",
 "car": {
  "modelName": "Audi",
  "color": "green"
 }
}
```

## Constructor parameters

Sometimes you don't really care or don't want to store some json property as a dedicated class field,
but instead, you would like to use it's value in constructor to calculate other class properties.
This way you don't have a convenience to annotate a class field, but you could utilize constructor parameter for that.

With the input JSON like this:
```json
{"LogistikTeileInOrdnung":"true"}
```

You could potentially have a class like this:
```dart
@jsonSerializable
class BusinessObject {
  final bool logisticsChecked;
  final bool logisticsOK;

  BusinessObject()
      : logisticsChecked = false,
        logisticsOK = true;

  @jsonConstructor
  BusinessObject.fromJson(
      @JsonProperty(name: 'LogistikTeileInOrdnung') String processed)
      : logisticsChecked = processed != null && processed != 'null',
        logisticsOK = processed == 'true';
}
```

## Unmapped properties

If you are looking for an alternative to Java Jackson `@JsonAnySetter / @JsonAnyGetter`
It is possible to configure the same scenario as follows:

```dart
@jsonSerializable
class UnmappedProperties {
  String name;

  Map<String, dynamic> _extraPropsMap = {};

  @jsonProperty
  void unmappedSet(String name, dynamic value) {
    _extraPropsMap[name] = value;
  }

  @jsonProperty
  Map<String, dynamic> unmappedGet() {
    return _extraPropsMap;
  }
}

// given
final json = '''{"name":"Bob","extra1":1,"extra2":"xxx"}''';

// when
final instance = JsonMapper.deserialize<UnmappedProperties>(json);

// then
expect(instance.name, 'Bob');
expect(instance._extraPropsMap['name'], null);
expect(instance._extraPropsMap['extra1'], 1);
expect(instance._extraPropsMap['extra2'], 'xxx');
```

## Iterable types

Since Dart language has no possibility to create typed iterables dynamically, it's a bit of a challenge
to create exact typed lists/sets/etc via reflection approach. Those types has to be declared explicitly.

For example List() will produce `List<dynamic>` type which can't be directly set to the concrete
target field `List<Car>` for instance. So obvious workaround will be to cast
`List<dynamic> => List<Car>`, which can be performed as `List<dynamic>().cast<Car>()`.

Basic iterable based generics using Dart built-in types like `List<num>, List<String>, List<bool>,

mapper/CHANGELOG.md
## 2.1.14

* #157, (feat) Serialize GET parameters object class to Uri. Brilliant basics

## 2.1.13

* generated adapter naming improved, lint warnings removed in generated code

## 2.1.12

* #159, (fix) Few minor fixes related to recursive objects processing

## 2.1.11

* #159, (fix) Proper support for nested parents in relative path reference to parent itself from nested object ".."

## 2.1.10

* #158, (fix) Local relative paths are not handled properly

## 2.1.9

* bump dart to `2.14.0`

## 2.1.8

* Remove local model packages from `dev_dependencies` to avoid penalties from pub.dev scoring algorithm

## 2.1.7

* #155, (feat) Introduce modularization, i.e. multi project setup is possible now. Model packages could be reused by different projects, flutter/web/etc.

## 2.1.6

* #150, (feat) Introduce `@JsonProperty.ignoreIfDefault` & `@Json.ignoreDefaultMembers`
* #152, (fix) Nullable list throws exception on being serialized when null

## 2.1.5

* (feat) Treat `ignoreForSerialization: true` by default when parent ".." is referenced OR field value is injected
* #147, (feat) Custom values injection during deserialization call

## 2.1.4

* #143, (fix) Unhandled Exception: type 'Null' is not a subtype of type 'Object'
* #144, (feat) Relative path reference to parent itself from nested object ".."

## 2.1.3

* replace `pedantic` linter by `lints` package
* Annotate annotations with the supported target types, to minimize incorrect usage.

## 2.1.2

* #137, Support custom `Iterable` converter
* bump dart to 2.13

## 2.1.1

* #131, `@JsonProperty` annotation in superclass should be inherited for overriden getters
* `deserialize` method extended to accept json `String` or parsed json `Object`
* #133, #135, second pass for converter during deserialization has been removed, more speed for deserialization

## 2.1.0

* #128, Introduce `flatten` attribute to `@JsonProperty` meta
* #97, sound null safety

## 2.0.9

* #127, Proper support for positional constructor arguments

## 2.0.8

* #126, Inaccurate implementation for contexts comparison improved

## 2.0.7

* #125, Fix null collection case with value decorator
* #120, Support read only getters as discriminator property

## 2.0.6

* #121, Support deserialization from type variable

## 2.0.5

* #120, Support for subclass using a discriminator property `@Json(typeNameProperty)` => `@Json(discriminatorProperty)`
        For migration from old versions just replace all `typeNameProperty` by `discriminatorProperty`.
        Global option `DeserializationOptions.typeNameProperty` has been removed.

## 2.0.4

* improve pub score for static analysis

## 2.0.3

* improve pub score for static analysis

## 2.0.2

* update changelog

## 2.0.1

* add readme

## 2.0.0

* #97, null safety support

## 1.7.13

* #118, fix: Add defaultValue handling for `enumConverterNumeric`
* #77, fix: Flutter for Web: Dynamic call with too many arguments. Arguments: [Instance of 'NativeJavaScriptObject']

## 1.7.12

* #117, fix: Cached values from converters made more accurate in terms of different types

## 1.7.11

* #115, fix: Null pointer on Flutter AOT app build
* #114, introduce async init method `initializeJsonMapperAsync`

## 1.7.10

* #111, feat: Support for relative reference to parent field from nested object "../id"
* #112, feat: Introduce support for field aliases. `@JsonProperty(name: ['alias', 'fullName', 'name'])`
* #77, fix: Flutter for Web: Dynamic call with too many arguments

## 1.7.9

* #110, feat: Support for Enum `EnumDescriptor.defultValue` on unknown values during deserialization

## 1.7.8

* #108, #109, feat: Flags `JsonProperty.notNull` `JsonProperty.required` spanned over serialization / deserialization process consistently
* #110, feat: Support for `String` / `Num` Enum values mappings introduced

## 1.7.7

* #106, fix: Do not generate invalid code when there is no main entry point provided, in case of library package for instance.
* #107, fix: Unmapped properties contains inherited readonly entries with custom json name
* #108, feat: Introduced `JsonProperty.required` to mark field as mandatory to be present in the JSON payload
* #109, feat: Introduced `JsonProperty.notNull` to mark field as mandatory to be present in the JSON payload & cannot be NULL

## 1.7.6

* #104, fix: Custom converter results cache layer does not work as expected

## 1.7.5

* #103, feat: Mixins Deserialization support introduced

## 1.7.4

* #100, fix: Naming conflicts in generated file. Prefixes introduced

## 1.7.3

* #99, fix: builder crawling over exports statements improved

## 1.7.2

* #91, Enum values inherit CaseStyle from the parent serializable class or global option.
* #94, fix: Deserialization fails when using [enumConverterNumeric]

## 1.7.1

* **Breaking change** Custom converters contract changed. `JsonProperty` replaced by serialization/deserialization `context`. Allowing each custom converter instance to have diverse context information
* #90, Support deserialization to default enum value for unknown values

## 1.7.0

* **Breaking change** Enums handling changed, no more repetitive `@JsonProperty(enumValues: Color.values)`. Annotate local Enums and Register third party Enums once per application
* Issue #88 addressed. Refine imports in auto-generated `*.mapper.g.dart` so no name conflicts will be possible

## 1.6.5

* Introduced converters for `RegExp`, `Uri` types.
* Internal `JsonMapper` methods made private.

mapper/example/example.mapper.g.dart
// This file has been generated by the naked_mapper package.
// https://github.com/k-paxian/dart-json-mapper
// @dart = 2.12
import 'example.dart' as x0 show Car, Person, Color;
import 'package:naked_mapper/dart_json_mapper.dart'
    show JsonMapper, JsonMapperAdapter, typeOf;
// This file has been generated by the reflectable package.
// https://github.com/dart-lang/reflectable.

import 'dart:core';
import 'example.dart' as prefix1;
import 'package:naked_mapper/src/model/annotations.dart' as prefix0;
import 'package:naked_mapper/src/model/converters.dart' as prefix2;

// ignore_for_file: prefer_adjacent_string_concatenation
// ignore_for_file: prefer_collection_literals
// ignore_for_file: unnecessary_const
// ignore_for_file: implementation_imports

// ignore:unused_import
import 'package:reflectable/mirrors.dart' as m;
// ignore:unused_import
import 'package:reflectable/src/reflectable_builder_based.dart' as r;
// ignore:unused_import
import 'package:reflectable/reflectable.dart' as r show Reflectable;

final _data = <r.Reflectable, r.ReflectorData>{
  const prefix0.JsonSerializable(): r.ReflectorData(
      <m.TypeMirror>[
        r.NonGenericClassMirrorImpl(
            r'Car',
            r'.Car',
            7,
            0,
            const prefix0.JsonSerializable(),
            const <int>[0, 1, 15, 20],
            const <int>[21, 15, 22, 23, 24, 16, 17, 18, 19],
            const <int>[],
            -1,
            {},
            {},
            {
              r'': (bool b) =>
                  (model, color) => b ? prefix1.Car(model, color) : null
            },
            -1,
            0,
            const <int>[],
            const [prefix0.jsonSerializable],
            null),
        r.NonGenericClassMirrorImpl(
            r'Person',
            r'.Person',
            7,
            1,
            const prefix0.JsonSerializable(),
            const <int>[2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 49, 50],
            const <int>[
              21,
              51,
              22,
              23,
              24,
              25,
              26,
              27,
              28,
              29,
              30,
              31,
              32,
              33,
              34,
              35,
              36,
              37,
              38,
              39,
              40,
              41,
              42,
              43,
              44,
              45,
              46,
              47,
              48,
              49
            ],
            const <int>[],
            -1,
            {},
            {},
            {r'': (bool b) => () => b ? prefix1.Person() : null},
            -1,
            1,
            const <int>[],
            const [prefix0.jsonSerializable],
            null),
        r.NonGenericClassMirrorImpl(
            r'Color',
            r'.Color',
            524295,
            2,
            const prefix0.JsonSerializable(),
            const <int>[52],
            const <int>[21, 52, 22, 23, 24, 53],
            const <int>[],
            -1,
            {},
            {},
            {},
            -1,
            2,
            const <int>[],
            const [prefix0.jsonSerializable],
            null)
      ],
      <m.DeclarationMirror>[
        r.VariableMirrorImpl(
            r'model',
            32773,
            0,
            const prefix0.JsonSerializable(),
            -1,
            3,
            3,
            const <int>[],
            const [const prefix0.JsonProperty(name: 'modelName')]),
        r.VariableMirrorImpl(r'color', 32773, 0,
            const prefix0.JsonSerializable(), 2, 2, 2, const <int>[], const []),
        r.VariableMirrorImpl(
            r'skills',
            2129925,
            1,
            const prefix0.JsonSerializable(),
            -1,
            4,
            5,
            const <int>[3],
            const []),
        r.VariableMirrorImpl(r'lastPromotionDate', 32773, 1,
            const prefix0.JsonSerializable(), -1, 6, 6, const <int>[], const [
          const prefix0.JsonProperty(name: 'last_promotion_date', ignore: true)
        ]),
        r.VariableMirrorImpl(
            r'hireDate',
            32773,
            1,
            const prefix0.JsonSerializable(),
            -1,
            7,
            7,
            const <int>[],
            const [const prefix0.JsonProperty(name: 'hire_date')]),
        r.VariableMirrorImpl(
            r'married',
            32773,
            1,
            const prefix0.JsonSerializable(),
            -1,
            8,
            8, const <int>[], const []),
        r.VariableMirrorImpl(
            r'name',
            32773,
            1,
            const prefix0.JsonSerializable(),
            -1,
            3,
            3, const <int>[], const []),
        r.VariableMirrorImpl(
            r'salary',
            32773,
            1,
            const prefix0.JsonSerializable(),
            -1,
            9,
            9,
            const <int>[],
            const [const prefix0.JsonProperty(ignore: true)]),
        r.VariableMirrorImpl(r'dob', 32773, 1, const prefix0.JsonSerializable(),
            -1, 9, 9, const <int>[], const []),
        r.VariableMirrorImpl(r'age', 32773, 1, const prefix0.JsonSerializable(),
            -1, 10, 10, const <int>[], const []),
        r.VariableMirrorImpl(
            r'lastName',
            32773,
            1,
            const prefix0.JsonSerializable(),
            -1,
            3,
            3, const <int>[], const []),
        r.VariableMirrorImpl(
            r'eyeColor',
            32773,
            1,
            const prefix0.JsonSerializable(),
            2,
            2,
            2,
            const <int>[],
            const [const prefix0.JsonProperty(name: 'eye_color')]),
        r.VariableMirrorImpl(r'hairColor', 32773, 1,
            const prefix0.JsonSerializable(), 2, 2, 2, const <int>[], const [
          const prefix0.JsonProperty(converter: prefix2.enumConverterNumeric)
        ]),
        r.VariableMirrorImpl(
            r'vehicles',
            2129925,
            1,
            const prefix0.JsonSerializable(),
            -1,
            11,
            12,
            const <int>[0],
            const []),
        r.VariableMirrorImpl(
            r'index',
            33861,
            2,
            const prefix0.JsonSerializable(),
            -1,
            13,
            13, const <int>[], const []),
        r.MethodMirrorImpl(r'toString', 131074, 0, -1, 3, 3, const <int>[],
            const <int>[], const prefix0.JsonSerializable(), const [override]),
        r.ImplicitGetterMirrorImpl(const prefix0.JsonSerializable(), 0, 16),
        r.ImplicitSetterMirrorImpl(const prefix0.JsonSerializable(), 0, 17),
        r.ImplicitGetterMirrorImpl(const prefix0.JsonSerializable(), 1, 18),
        r.ImplicitSetterMirrorImpl(const prefix0.JsonSerializable(), 1, 19),
        r.MethodMirrorImpl(r'', 0, 0, -1, 0, 0, const <int>[],
            const <int>[0, 1], const prefix0.JsonSerializable(), const []),
        r.MethodMirrorImpl(r'==', 131074, -1, -1, 8, 8, const <int>[],
            const <int>[4], const prefix0.JsonSerializable(), const []),
        r.MethodMirrorImpl(
            r'noSuchMethod',
            65538,
            -1,
            -1,
            -1,
            -1,
            const <int>[],
            const <int>[5],
            const prefix0.JsonSerializable(),
            const []),
        r.MethodMirrorImpl(r'hashCode', 131075, -1, -1, 13, 13, const <int>[],
            const <int>[], const prefix0.JsonSerializable(), const []),
        r.MethodMirrorImpl(
            r'runtimeType',
            131075,
            -1,
            -1,
            14,
            14,
            const <int>[],
            const <int>[],
            const prefix0.JsonSerializable(),
            const []),
        r.ImplicitGetterMirrorImpl(const prefix0.JsonSerializable(), 2, 25),
        r.ImplicitSetterMirrorImpl(const prefix0.JsonSerializable(), 2, 26),
        r.ImplicitGetterMirrorImpl(const prefix0.JsonSerializable(), 3, 27),
        r.ImplicitSetterMirrorImpl(const prefix0.JsonSerializable(), 3, 28),
        r.ImplicitGetterMirrorImpl(const prefix0.JsonSerializable(), 4, 29),
        r.ImplicitSetterMirrorImpl(const prefix0.JsonSerializable(), 4, 30),
        r.ImplicitGetterMirrorImpl(const prefix0.JsonSerializable(), 5, 31),
        r.ImplicitSetterMirrorImpl(const prefix0.JsonSerializable(), 5, 32),
        r.ImplicitGetterMirrorImpl(const prefix0.JsonSerializable(), 6, 33),
        r.ImplicitSetterMirrorImpl(const prefix0.JsonSerializable(), 6, 34),
        r.ImplicitGetterMirrorImpl(const prefix0.JsonSerializable(), 7, 35),
        r.ImplicitSetterMirrorImpl(const prefix0.JsonSerializable(), 7, 36),
        r.ImplicitGetterMirrorImpl(const prefix0.JsonSerializable(), 8, 37),
        r.ImplicitSetterMirrorImpl(const prefix0.JsonSerializable(), 8, 38),
        r.ImplicitGetterMirrorImpl(const prefix0.JsonSerializable(), 9, 39),
        r.ImplicitSetterMirrorImpl(const prefix0.JsonSerializable(), 9, 40),
        r.ImplicitGetterMirrorImpl(const prefix0.JsonSerializable(), 10, 41),
        r.ImplicitSetterMirrorImpl(const prefix0.JsonSerializable(), 10, 42),
        r.ImplicitGetterMirrorImpl(const prefix0.JsonSerializable(), 11, 43),
        r.ImplicitSetterMirrorImpl(const prefix0.JsonSerializable(), 11, 44),
        r.ImplicitGetterMirrorImpl(const prefix0.JsonSerializable(), 12, 45),
        r.ImplicitSetterMirrorImpl(const prefix0.JsonSerializable(), 12, 46),
        r.ImplicitGetterMirrorImpl(const prefix0.JsonSerializable(), 13, 47),
        r.ImplicitSetterMirrorImpl(const prefix0.JsonSerializable(), 13, 48),
        r.MethodMirrorImpl(r'fullName', 131075, 1, -1, 3, 3, const <int>[],
            const <int>[], const prefix0.JsonSerializable(), const []),
        r.MethodMirrorImpl(r'', 64, 1, -1, 1, 1, const <int>[], const <int>[],
            const prefix0.JsonSerializable(), const []),
        r.MethodMirrorImpl(r'toString', 131074, -1, -1, 3, 3, const <int>[],
            const <int>[], const prefix0.JsonSerializable(), const []),
        r.MethodMirrorImpl(r'toString', 131138, 2, -1, 3, 3, const <int>[],
            const <int>[], const prefix0.JsonSerializable(), const []),
        r.ImplicitGetterMirrorImpl(const prefix0.JsonSerializable(), 14, 53)
      ],
      <m.ParameterMirror>[
        r.ParameterMirrorImpl(
            r'model',
            33798,
            20,
            const prefix0.JsonSerializable(),
            -1,
            3,
            3,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'color',
            33798,
            20,
            const prefix0.JsonSerializable(),
            2,
            2,
            2,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'_model',
            32870,
            17,
            const prefix0.JsonSerializable(),
            -1,
            3,
            3,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'_color',
            32870,
            19,
            const prefix0.JsonSerializable(),
            2,
            2,
            2,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'other',
            32774,
            21,
            const prefix0.JsonSerializable(),
            -1,
            15,
            15,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'invocation',
            32774,
            22,
            const prefix0.JsonSerializable(),
            -1,
            16,
            16,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'_skills',
            2130022,
            26,
            const prefix0.JsonSerializable(),
            -1,
            4,
            5,
            const <int>[3],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'_lastPromotionDate',
            32870,
            28,
            const prefix0.JsonSerializable(),
            -1,
            6,
            6,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'_hireDate',
            32870,
            30,
            const prefix0.JsonSerializable(),
            -1,
            7,
            7,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'_married',
            32870,
            32,
            const prefix0.JsonSerializable(),
            -1,
            8,
            8,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'_name',
            32870,
            34,
            const prefix0.JsonSerializable(),
            -1,
            3,
            3,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'_salary',
            32870,
            36,
            const prefix0.JsonSerializable(),
            -1,
            9,
            9,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'_dob',
            32870,
            38,
            const prefix0.JsonSerializable(),
            -1,
            9,
            9,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'_age',
            32870,
            40,
            const prefix0.JsonSerializable(),
            -1,
            10,
            10,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'_lastName',
            32870,
            42,
            const prefix0.JsonSerializable(),
            -1,
            3,
            3,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'_eyeColor',
            32870,
            44,
            const prefix0.JsonSerializable(),
            2,
            2,
            2,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'_hairColor',
            32870,
            46,
            const prefix0.JsonSerializable(),
            2,
            2,
            2,
            const <int>[],
            const [],
            null,
            null),
        r.ParameterMirrorImpl(
            r'_vehicles',
            2130022,
            48,
            const prefix0.JsonSerializable(),
            -1,
            11,
            12,
            const <int>[0],
            const [],
            null,
            null)
      ],
      <Type>[
        prefix1.Car,
        prefix1.Person,
        prefix1.Color,
        String,
        const m.TypeValue<List<String>>().type,
        List,
        DateTime,
        DateTime,
        bool,
        num,
        num,
        const m.TypeValue<List<prefix1.Car>>().type,
        List,
        int,
        Type,
        Object,
        Invocation
      ],
      3,
      {
        r'==': (dynamic instance) => (x) => instance == x,
        r'toString': (dynamic instance) => instance.toString,
        r'noSuchMethod': (dynamic instance) => instance.noSuchMethod,
        r'hashCode': (dynamic instance) => instance.hashCode,
        r'runtimeType': (dynamic instance) => instance.runtimeType,
        r'model': (dynamic instance) => instance.model,
        r'color': (dynamic instance) => instance.color,
        r'skills': (dynamic instance) => instance.skills,
        r'lastPromotionDate': (dynamic instance) => instance.lastPromotionDate,
        r'hireDate': (dynamic instance) => instance.hireDate,
        r'married': (dynamic instance) => instance.married,
        r'name': (dynamic instance) => instance.name,
        r'salary': (dynamic instance) => instance.salary,
        r'dob': (dynamic instance) => instance.dob,
        r'age': (dynamic instance) => instance.age,
        r'lastName': (dynamic instance) => instance.lastName,
        r'eyeColor': (dynamic instance) => instance.eyeColor,
        r'hairColor': (dynamic instance) => instance.hairColor,
        r'vehicles': (dynamic instance) => instance.vehicles,
        r'fullName': (dynamic instance) => instance.fullName,
        r'index': (dynamic instance) => instance.index
      },
      {
        r'model=': (dynamic instance, value) => instance.model = value,
        r'color=': (dynamic instance, value) => instance.color = value,
        r'skills=': (dynamic instance, value) => instance.skills = value,
        r'lastPromotionDate=': (dynamic instance, value) =>
            instance.lastPromotionDate = value,
        r'hireDate=': (dynamic instance, value) => instance.hireDate = value,
        r'married=': (dynamic instance, value) => instance.married = value,
        r'name=': (dynamic instance, value) => instance.name = value,
        r'salary=': (dynamic instance, value) => instance.salary = value,
        r'dob=': (dynamic instance, value) => instance.dob = value,
        r'age=': (dynamic instance, value) => instance.age = value,
        r'lastName=': (dynamic instance, value) => instance.lastName = value,
        r'eyeColor=': (dynamic instance, value) => instance.eyeColor = value,
        r'hairColor=': (dynamic instance, value) => instance.hairColor = value,
        r'vehicles=': (dynamic instance, value) => instance.vehicles = value
      },
      null,
      [])
};

final _memberSymbolMap = null;

void _initializeReflectable(JsonMapperAdapter adapter) {
  if (!adapter.isGenerated) {
    return;
  }
  r.data = adapter.reflectableData!;
  r.memberSymbolMap = adapter.memberSymbolMap;
}

final exampleGeneratedAdapter = JsonMapperAdapter(
    title: 'Example Generated Adapter',
    url: 'asset:naked_mapper/example/example.dart',
    reflectableData: _data,
    memberSymbolMap: _memberSymbolMap,
    valueDecorators: {
      typeOf<List<x0.Car>>(): (value) => value.cast<x0.Car>(),
      typeOf<Set<x0.Car>>(): (value) => value.cast<x0.Car>(),
      typeOf<List<x0.Person>>(): (value) => value.cast<x0.Person>(),
      typeOf<Set<x0.Person>>(): (value) => value.cast<x0.Person>(),
      typeOf<List<x0.Color>>(): (value) => value.cast<x0.Color>(),
      typeOf<Set<x0.Color>>(): (value) => value.cast<x0.Color>()
    },
    enumValues: {
      x0.Color: x0.Color.values
    });

Future<JsonMapper> initializeJsonMapperAsync(
        {Iterable<JsonMapperAdapter> adapters = const []}) =>
    Future(() => initializeJsonMapper(adapters: adapters));

JsonMapper initializeJsonMapper(
    {Iterable<JsonMapperAdapter> adapters = const []}) {
  JsonMapper.enumerateAdapters([...adapters, exampleGeneratedAdapter],
      (JsonMapperAdapter adapter) {
    _initializeReflectable(adapter);
    JsonMapper().useAdapter(adapter);
  });
  return JsonMapper();
}
